pkg/auth/sync/secret.go

```diff
@@ -0,0 +1,129 @@
+package sync
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/rancher/harvester/pkg/settings"
+	ctlcorev1 "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+
+	dashboardsync "github.com/kubernetes/dashboard/src/app/backend/sync/api"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+func NewSecretSynchronizer(secrets ctlcorev1.SecretClient, namespace, name string) *SecretSynchronizer {
+	return &SecretSynchronizer{
+		secrets:   secrets,
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+// Implements kubernetes dashboard Synchronizer interface.
+type SecretSynchronizer struct {
+	secret    *v1.Secret
+	secrets   ctlcorev1.SecretClient
+	name      string
+	namespace string
+
+	mux sync.Mutex
+}
+
+func (s *SecretSynchronizer) Name() string {
+	return fmt.Sprintf("%s-%s", s.name, s.namespace)
+}
+
+func (s *SecretSynchronizer) Create(obj runtime.Object) error {
+	secret, err := s.getSecret(obj)
+	if err != nil {
+		return err
+	}
+
+	_, err = s.secrets.Create(secret)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SecretSynchronizer) Get() runtime.Object {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	if s.secret == nil {
+		// In case secret was not yet initialized try to do it synchronously
+		secret, err := s.secrets.Get(s.namespace, s.name, metaV1.GetOptions{})
+		if err != nil {
+			return nil
```

Comment:
Why do we want to ignore this

Reply:
This SecretSynchronizer Get is an implement of the interface [Synchronizer](https://github.com/kubernetes/dashboard/blob/9306c63e05b085458d0adb8a537fbd1dc742ae8a/src/app/backend/sync/api/types.go#L29) and be used [here](https://github.com/kubernetes/dashboard/blob/9306c63e05b085458d0adb8a537fbd1dc742ae8a/src/app/backend/auth/jwe/keyholder.go#L125) , when return nil, it will create a new one, the nil return is for init

Reply:
In this implementation if an error occurs there's no guarantee that it is because the object is not created, which seems contradictory to the interface definition. I think we might want to cache the secret on setup, not in the Get() implementation.

Reply:
I agree if we can return error or panic when can't find the secret, but we use some function from k8s dashboard, for example [here](https://github.com/kubernetes/dashboard/blob/master/src/app/backend/auth/jwe/keyholder.go#L169), it will init the secret, in the init function if it get nil return from the Get(), will call Create(), it will still [panic](https://github.com/kubernetes/dashboard/blob/9306c63e05b085458d0adb8a537fbd1dc742ae8a/src/app/backend/auth/jwe/keyholder.go#L135) when the creation is failed.

Reply:
add more comment for handle panic and update Refresh, panic when can't find secret